<a href="https://colab.research.google.com/github/congtuan99/learning-data/blob/main/day3_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Block 1: Pandas

In [5]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [6]:
df = sns.load_dataset("penguins")

In [7]:
df.head()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,Male
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,Female
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,Female
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,Female


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


In [9]:

df.isna()
# de même
df.isnull()

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
0,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False
3,False,False,True,True,True,True,True
4,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...
339,False,False,True,True,True,True,True
340,False,False,False,False,False,False,False
341,False,False,False,False,False,False,False
342,False,False,False,False,False,False,False


In [10]:
df.isnull().sum()


,0
species,0
island,0
bill_length_mm,2
bill_depth_mm,2
flipper_length_mm,2
body_mass_g,2
sex,11


In [11]:

df[df["body_mass_g"].isnull()]



,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN
339,Gentoo,Biscoe,NaN,NaN,NaN,NaN,NaN


In [34]:

df["body_mass_g"].fillna(df["body_mass_g"].mean())




,body_mass_g
0,3750.000000
1,3800.000000
2,3250.000000
3,4201.754386
4,3450.000000
...,...
339,4201.754386
340,4850.000000
341,5750.000000
342,5200.000000


In [35]:

df["sex"].fillna("Male") # exemple

,sex
0,Male
1,Female
2,Female
3,Male
4,Female
...,...
339,Male
340,Female
341,Male
342,Female


In [17]:
df[df["sex"].isna()]

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN
8,Adelie,Torgersen,34.1,18.1,193.0,3475.0,NaN
9,Adelie,Torgersen,42.0,20.2,190.0,4250.0,NaN
10,Adelie,Torgersen,37.8,17.1,186.0,3300.0,NaN
11,Adelie,Torgersen,37.8,17.3,180.0,3700.0,NaN
47,Adelie,Dream,37.5,18.9,179.0,2975.0,NaN
246,Gentoo,Biscoe,44.5,14.3,216.0,4100.0,NaN
286,Gentoo,Biscoe,46.2,14.4,214.0,4650.0,NaN
324,Gentoo,Biscoe,47.3,13.8,216.0,4725.0,NaN
336,Gentoo,Biscoe,44.5,15.7,217.0,4875.0,NaN


In [49]:
df_clean1 = df.dropna()
print(df_clean1)
print(df_clean1.shape,df.shape)

    species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
4    Adelie  Torgersen            36.7           19.3              193.0   
5    Adelie  Torgersen            39.3           20.6              190.0   
..      ...        ...             ...            ...                ...   
338  Gentoo     Biscoe            47.2           13.7              214.0   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         3750.0    Male  
1         3800.0  Female  
2     

In [19]:
df_clean2 = df.copy()
df_clean2["body_mass_g"] = df["body_mass_g"].fillna(df["body_mass_g"].median())
df_clean2["body_mass_g"]

,body_mass_g
0,3750.0
1,3800.0
2,3250.0
3,4050.0
4,3450.0
...,...
339,4050.0
340,4850.0
341,5750.0
342,5200.0


Commentaire: on remplace les valeurs manquantes NaN par la médiane  de la colonne au lieu de la moyenne. Cela rend le résultat plus robuste pour les analyses ultérieures parce que la médiane est moins sensible aux valeurs extrêmes (aberrantes) que la moyenne.


# BLOCK 2: SQL

In [20]:
import duckdb

In [37]:
duckdb.sql("select count(*), count(body_mass_g) from df")


┌──────────────┬────────────────────┐
│ count_star() │ count(body_mass_g) │
│    int64     │       int64        │
├──────────────┼────────────────────┤
│          344 │                342 │
└──────────────┴────────────────────┘

In [38]:

duckdb.sql("select * from df where sex is null")

┌─────────┬───────────┬────────────────┬───────────────┬───────────────────┬─────────────┬─────────┐
│ species │  island   │ bill_length_mm │ bill_depth_mm │ flipper_length_mm │ body_mass_g │   sex   │
│ varchar │  varchar  │     double     │    double     │      double       │   double    │ varchar │
├─────────┼───────────┼────────────────┼───────────────┼───────────────────┼─────────────┼─────────┤
│ Adelie  │ Torgersen │           NULL │          NULL │              NULL │        NULL │ NULL    │
│ Adelie  │ Torgersen │           34.1 │          18.1 │             193.0 │      3475.0 │ NULL    │
│ Adelie  │ Torgersen │           42.0 │          20.2 │             190.0 │      4250.0 │ NULL    │
│ Adelie  │ Torgersen │           37.8 │          17.1 │             186.0 │      3300.0 │ NULL    │
│ Adelie  │ Torgersen │           37.8 │          17.3 │             180.0 │      3700.0 │ NULL    │
│ Adelie  │ Dream     │           37.5 │          18.9 │             179.0 │      2975.0 │ 

# Block 3:  MERGE et JOIN

In [39]:
# on va créer un autre table: nom scientifique des composition des espèces
especes = pd.DataFrame({
    "species" : ["Adelie","Gentoo","Chinstrap"],
    "nom_scientifique" : ["Pygoscelis adeliae", "Pygoscelis papua", "Pygoscelis antarcticus"]
})

especes

,species,nom_scientifique
0,Adelie,Pygoscelis adeliae
1,Gentoo,Pygoscelis papua
2,Chinstrap,Pygoscelis antarcticus


In [40]:

duckdb.sql("select * from df join especes on df.species=especes.species")


┌─────────┬───────────┬────────────────┬───────────────┬───────────────────┬─────────────┬─────────┬─────────┬────────────────────┐
│ species │  island   │ bill_length_mm │ bill_depth_mm │ flipper_length_mm │ body_mass_g │   sex   │ species │  nom_scientifique  │
│ varchar │  varchar  │     double     │    double     │      double       │   double    │ varchar │ varchar │      varchar       │
├─────────┼───────────┼────────────────┼───────────────┼───────────────────┼─────────────┼─────────┼─────────┼────────────────────┤
│ Adelie  │ Torgersen │           39.1 │          18.7 │             181.0 │      3750.0 │ Male    │ Adelie  │ Pygoscelis adeliae │
│ Adelie  │ Torgersen │           39.5 │          17.4 │             186.0 │      3800.0 │ Female  │ Adelie  │ Pygoscelis adeliae │
│ Adelie  │ Torgersen │           40.3 │          18.0 │             195.0 │      3250.0 │ Female  │ Adelie  │ Pygoscelis adeliae │
│ Adelie  │ Torgersen │           NULL │          NULL │              NULL │

In [45]:


especes_nan = pd.DataFrame({
    "species" : ["Adelie", "Gentoo", "Chinstrap"],
    "nom_scientifique" : ["Pygoscelis adeliae", "Pygoscelis papua", np.nan ] # để 1 dòng có giá trị NaN => dùng np.nan hoặc None
})
especes_nan

,species,nom_scientifique
0,Adelie,Pygoscelis adeliae
1,Gentoo,Pygoscelis papua
2,Chinstrap,NaN


In [46]:

df_merge1 = df.merge(especes_nan, on ="species", how="left")  # Pandas
print(df_merge1.isna().sum())
df_merge1[df_merge1["species"]=="Chinstrap" ]

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
nom_scientifique     68
dtype: int64


,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex,nom_scientifique
152,Chinstrap,Dream,46.5,17.9,192.0,3500.0,Female,NaN
153,Chinstrap,Dream,50.0,19.5,196.0,3900.0,Male,NaN
154,Chinstrap,Dream,51.3,19.2,193.0,3650.0,Male,NaN
155,Chinstrap,Dream,45.4,18.7,188.0,3525.0,Female,NaN
156,Chinstrap,Dream,52.7,19.8,197.0,3725.0,Male,NaN
...,...,...,...,...,...,...,...,...
215,Chinstrap,Dream,55.8,19.8,207.0,4000.0,Male,NaN
216,Chinstrap,Dream,43.5,18.1,202.0,3400.0,Female,NaN
217,Chinstrap,Dream,49.6,18.2,193.0,3775.0,Male,NaN
218,Chinstrap,Dream,50.8,19.0,210.0,4100.0,Male,NaN


In [43]:
# SQL
duckdb.sql("select count(*) as nb_null from df left join especes1 on df.species = especes1.species where especes1.nom_scientifique is null")

┌─────────┐
│ nb_null │
│  int64  │
├─────────┤
│      68 │
└─────────┘

In [44]:

duckdb.sql("select * from df left join especes1 e1 on df.species=e1.species where e1.nom_scientifique is null")

┌───────────┬─────────┬────────────────┬───────────────┬───────────────────┬─────────────┬─────────┬───────────┬──────────────────┐
│  species  │ island  │ bill_length_mm │ bill_depth_mm │ flipper_length_mm │ body_mass_g │   sex   │  species  │ nom_scientifique │
│  varchar  │ varchar │     double     │    double     │      double       │   double    │ varchar │  varchar  │     varchar      │
├───────────┼─────────┼────────────────┼───────────────┼───────────────────┼─────────────┼─────────┼───────────┼──────────────────┤
│ Chinstrap │ Dream   │           46.5 │          17.9 │             192.0 │      3500.0 │ Female  │ Chinstrap │ NULL             │
│ Chinstrap │ Dream   │           50.0 │          19.5 │             196.0 │      3900.0 │ Male    │ Chinstrap │ NULL             │
│ Chinstrap │ Dream   │           51.3 │          19.2 │             193.0 │      3650.0 │ Male    │ Chinstrap │ NULL             │
│ Chinstrap │ Dream   │           45.4 │          18.7 │             188.0 │

In [30]:

df_merge = df.merge(especes,on="species")
df_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
 7   nom_scientifique   344 non-null    object 
dtypes: float64(4), object(4)
memory usage: 21.6+ KB


In [31]:
df_merge

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex,nom_scientifique
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,Male,Pygoscelis adeliae
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,Female,Pygoscelis adeliae
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,Female,Pygoscelis adeliae
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN,Pygoscelis adeliae
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,Female,Pygoscelis adeliae
...,...,...,...,...,...,...,...,...
339,Gentoo,Biscoe,NaN,NaN,NaN,NaN,NaN,Pygoscelis papua
340,Gentoo,Biscoe,46.8,14.3,215.0,4850.0,Female,Pygoscelis papua
341,Gentoo,Biscoe,50.4,15.7,222.0,5750.0,Male,Pygoscelis papua
342,Gentoo,Biscoe,45.2,14.8,212.0,5200.0,Female,Pygoscelis papua


In [47]:
especes_sans_chinstrap = pd.DataFrame({
    "species" : ["Adelie", "Gentoo"],
    "nom_scientifique" : ["Pygoscelis adeliae", "Pygoscelis papua"]
})
merge = df.merge(especes_sans_chinstrap, on="species")
mergeleft = df.merge(especes_sans_chinstrap, on="species",how="left")
print(merge.shape, mergeleft.shape)

(276, 8) (344, 8)


 Commentaire: à partir du résultat, le nombre de lignes de la table obtenue par inner join est 276, alors que le nombre de lignes de la table obtenu par left join est 344. Cela s'explique par la manière dont la jointure fonctionne. L'inner join supprime automatiquement les lignes de la table de gauche qui n'ont pas de correspondance dans la table de droite. En revanche, left join conserve toutes les lignes de table de gauche, même sans correspondance à droite. Les colonnes manquantes sont alors remplies par NaN.